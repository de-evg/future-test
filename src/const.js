export const showingStatus = {
    UNSET: `UNSET`,
    SHOW: `SHOW`,
    HIDE: `HIDE`
};

export const DEFAULT_SORT = `ID_UP`;

export const SortType = {
    ID_UP: `ID_UP`,
    ID_DOWN: `ID_DOWN`,
    FIRSTNAME_UP: `FIRSTNAME_UP`,
    FIRSTNAME_DOWN: `FIRSTNAME_DOWN`,
    LASTNAME_UP: `LASTNAME_UP`,
    LASTNAME_DOWN: `LASTNAME_DOWN`,
    EMAIL_UP: `EMAIL_UP`,
    EMAIL_DOWN: `EMAIL_DOWN`,
    PHONE_UP: `PHONE_UP`,
    PHONE_DOWN: `PHONE_DOWN`,
};