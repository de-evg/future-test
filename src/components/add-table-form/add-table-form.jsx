import React, {useCallback} from 'react';
import PropTypes from 'prop-types';
import Table from '../table/table';
import AddTableBody from "../add-table-body/add-table-body";
import {connect} from 'react-redux';
import {NameSpace} from '../../store/reducers/root';
import {showingStatus} from '../../const';
import {ActionCreator} from "../../store/action";

const AddTableForm = ({
  addRowFormVisualStatus, 
  submitNewRow, 
  resetInputs, 
  id, 
  firstName, 
  lastName, 
  email, 
  phone, 
  isIdValid,
  isFirstNameValid,
  isLastNameValid,
  isPhoneValid,
  isEmailValid}) => {
    
  let visualClass = "";
  if (addRowFormVisualStatus === showingStatus.HIDE) {
    visualClass = "add-row-form--hidden";
  }
  if (addRowFormVisualStatus === showingStatus.SHOW) {
    visualClass = "add-row-form--show";
  }

  const handleSubmit = useCallback((evt) => {
    evt.preventDefault();
    if (isIdValid && isFirstNameValid && isLastNameValid && isPhoneValid && isEmailValid) {
      submitNewRow({
        id, firstName, lastName, email, phone
      });
      resetInputs();
    }    
  }, [submitNewRow, resetInputs, id, firstName, lastName, email, phone,isIdValid, isFirstNameValid, isLastNameValid, isPhoneValid, isEmailValid]);

  return (
    <form onSubmit={handleSubmit} action="#" className={`add-row-form ${visualClass}`}>
      <Table tableBody={<AddTableBody />} />
      <button type="submit" className="add-row-form__button"></button>
    </form>
  );
};

AddTableForm.propTypes = {
  addRowFormVisualStatus: PropTypes.string.isRequired,
  resetInputs: PropTypes.func.isRequired,
  submitNewRow: PropTypes.func.isRequired,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  firstName: PropTypes.string.isRequired,
  lastName: PropTypes.string.isRequired,
  email: PropTypes.string.isRequired,
  phone: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
  isIdValid: PropTypes.bool.isRequired,
  isFirstNameValid: PropTypes.bool.isRequired,
  isLastNameValid: PropTypes.bool.isRequired,
  isPhoneValid: PropTypes.bool.isRequired,
  isEmailValid: PropTypes.bool.isRequired,
};

const mapStateToProps = (state) => ({
  addRowFormVisualStatus: state[NameSpace.INTERFACE_DATA].addRowFormVisualStatus,
  id: state[NameSpace.NEW_ROW].id,
  firstName: state[NameSpace.NEW_ROW].firstName,
  lastName: state[NameSpace.NEW_ROW].lastName,
  phone: state[NameSpace.NEW_ROW].phone,
  email: state[NameSpace.NEW_ROW].email,
  isIdValid: state[NameSpace.NEW_ROW].isIdValid,
  isFirstNameValid: state[NameSpace.NEW_ROW].isFirstNameValid,
  isLastNameValid: state[NameSpace.NEW_ROW].isLastNameValid,
  isPhoneValid: state[NameSpace.NEW_ROW].isPhoneValid,
  isEmailValid: state[NameSpace.NEW_ROW].isEmailValid,
});

const mapDispatchToProps = (dispatch) => ({
  submitNewRow(newRow) {
    dispatch(ActionCreator.submitNewRow(newRow))
  },
  resetInputs() {
    dispatch(ActionCreator.resetNewRow())
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTableForm);
