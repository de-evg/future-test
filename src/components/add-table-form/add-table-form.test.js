import React from 'react';
import renderer from "react-test-renderer";
import {showingStatus} from '../../const';
import {AddTableForm} from "./add-table-form";

describe(`Render AddTableForm`, () => {
  it(`Should AddTableForm render correctly with UNSET showing status`, () => {
    const tree = renderer
      .create(
        <AddTableForm
          addRowFormVisualStatus={showingStatus.UNSET}
          submitNewRow={() => {}}
            />, {
          createNodeMock: () => {
            return {};
          }
        }
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });

  it(`Should AddTableForm render correctly with SHOW showing status`, () => {
    const tree = renderer
      .create(
        <AddTableForm
          addRowFormVisualStatus={showingStatus.SHOW}
          submitNewRow={() => {}}
            />, {
          createNodeMock: () => {
            return {};
          }
        }
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });
  
  it(`Should AddTableForm render correctly with HIDE showing status`, () => {
    const tree = renderer
      .create(
        <AddTableForm
          addRowFormVisualStatus={showingStatus.HIDE}
          submitNewRow={() => {}}
            />, {
          createNodeMock: () => {
            return {};
          }
        }
      )
      .toJSON();

      expect(tree).toMatchSnapshot();
  });
});
