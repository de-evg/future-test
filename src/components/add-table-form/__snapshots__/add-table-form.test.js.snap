// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render AddTableForm Should AddTableForm render correctly with HIDE showing status 1`] = `
<form
  action="#"
  className="add-row-form add-row-form--hidden"
  onSubmit={[Function]}
>
  <table
    className="table"
  >
    <thead>
      <tr
        className="table__header table__row"
      >
        <th
          className="table__column-title table__data"
        >
          ID
        </th>
        <th
          className="table__column-title table__data"
        >
          First Name
        </th>
        <th
          className="table__column-title table__data"
        >
          Last Name
        </th>
        <th
          className="table__column-title table__data"
        >
          Email
        </th>
        <th
          className="table__column-title table__data"
        >
          Phone
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="table__row"
      >
        <td
          className="table__data"
        >
          <input
            id="id"
            name="id"
            onChange={[Function]}
            placeholder="ID"
            required={true}
            type="number"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="firstName"
            name="firstName"
            onChange={[Function]}
            placeholder="Имя"
            required={true}
            type="text"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="lastName"
            name="lastName"
            onChange={[Function]}
            placeholder="Фамилия"
            required={true}
            type="text"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="your@email.ru"
            required={true}
            type="email"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="phone"
            name="phone"
            onChange={[Function]}
            placeholder="+7 999 99 99"
            required={true}
            type="tel"
            value=""
          />
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="add-row-form__button"
    type="submit"
  />
</form>
`;

exports[`Render AddTableForm Should AddTableForm render correctly with SHOW showing status 1`] = `
<form
  action="#"
  className="add-row-form add-row-form--show"
  onSubmit={[Function]}
>
  <table
    className="table"
  >
    <thead>
      <tr
        className="table__header table__row"
      >
        <th
          className="table__column-title table__data"
        >
          ID
        </th>
        <th
          className="table__column-title table__data"
        >
          First Name
        </th>
        <th
          className="table__column-title table__data"
        >
          Last Name
        </th>
        <th
          className="table__column-title table__data"
        >
          Email
        </th>
        <th
          className="table__column-title table__data"
        >
          Phone
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="table__row"
      >
        <td
          className="table__data"
        >
          <input
            id="id"
            name="id"
            onChange={[Function]}
            placeholder="ID"
            required={true}
            type="number"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="firstName"
            name="firstName"
            onChange={[Function]}
            placeholder="Имя"
            required={true}
            type="text"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="lastName"
            name="lastName"
            onChange={[Function]}
            placeholder="Фамилия"
            required={true}
            type="text"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="your@email.ru"
            required={true}
            type="email"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="phone"
            name="phone"
            onChange={[Function]}
            placeholder="+7 999 99 99"
            required={true}
            type="tel"
            value=""
          />
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="add-row-form__button"
    type="submit"
  />
</form>
`;

exports[`Render AddTableForm Should AddTableForm render correctly with UNSET showing status 1`] = `
<form
  action="#"
  className="add-row-form "
  onSubmit={[Function]}
>
  <table
    className="table"
  >
    <thead>
      <tr
        className="table__header table__row"
      >
        <th
          className="table__column-title table__data"
        >
          ID
        </th>
        <th
          className="table__column-title table__data"
        >
          First Name
        </th>
        <th
          className="table__column-title table__data"
        >
          Last Name
        </th>
        <th
          className="table__column-title table__data"
        >
          Email
        </th>
        <th
          className="table__column-title table__data"
        >
          Phone
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="table__row"
      >
        <td
          className="table__data"
        >
          <input
            id="id"
            name="id"
            onChange={[Function]}
            placeholder="ID"
            required={true}
            type="number"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="firstName"
            name="firstName"
            onChange={[Function]}
            placeholder="Имя"
            required={true}
            type="text"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="lastName"
            name="lastName"
            onChange={[Function]}
            placeholder="Фамилия"
            required={true}
            type="text"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="email"
            name="email"
            onChange={[Function]}
            placeholder="your@email.ru"
            required={true}
            type="email"
            value=""
          />
        </td>
        <td
          className="table__data"
        >
          <input
            id="phone"
            name="phone"
            onChange={[Function]}
            placeholder="+7 999 99 99"
            required={true}
            type="tel"
            value=""
          />
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="add-row-form__button"
    type="submit"
  />
</form>
`;
